### ISSUE TITLE
Enable Mongo API versioning

### ISSUE THREAD

**Issue by pricehatfield** [hashicorp/terraform-provider-azurerm#10963]
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Description

Allow versioning of the Mongo API when creating the Cosmos DB account.  4.0 is now available in Azure, but not through this provider.

Prior endeavors have been made to expose versions through alternate "capabilities" values, but it would be cleaner to move this to its own node and effectively make it a pass-through value for Azure to validate.  Per documentation, this "server version" only applies for the Mongo API.

### References

Prior work on 3.4 vs. 3.6 versions:
* https://github.com/terraform-providers/terraform-provider-azurerm/issues/4757

Documentation on 4.0:
* https://docs.microsoft.com/en-us/azure/cosmos-db/mongodb-feature-support-40
* https://docs.microsoft.com/en-us/cli/azure/cosmosdb?view=azure-cli-latest#az_cosmosdb_create
* Sample ARM template downloaded from Azure uses the "apiProperties" peer of "capabilities":
<pre>
{
    "$schema": "http://schema.management.azure.com/schemas/2015-01-01/deploymentTemplate.json#",
    "contentVersion": "1.0.0.0",
    "parameters": {
        "name": {
            "type": "string"
        },
        "location": {
            "type": "string"
        },
        "locationName": {
            "type": "string"
        },
        "defaultExperience": {
            "type": "string"
        },
        "isZoneRedundant": {
            "type": "string"
        }
    },
    "variables": {},
    "resources": [
        {
            "name": "[concat(parameters('name'), '/default')]",
            "type": "Microsoft.DocumentDB/databaseAccounts/notebookWorkspaces",
            "properties": {},
            "dependsOn": [
                "[resourceId('Microsoft.DocumentDB/databaseAccounts', parameters('name'))]"
            ],
            "apiVersion": "2020-04-01"
        },
        {
            "apiVersion": "2020-04-01",
            "kind": "MongoDB",
            "type": "Microsoft.DocumentDb/databaseAccounts",
            "name": "[parameters('name')]",
            "location": "[parameters('location')]",
            "properties": {
                "databaseAccountOfferType": "Standard",
                "locations": [
                    {
                        "id": "[concat(parameters('name'), '-', parameters('location'))]",
                        "failoverPriority": 0,
                        "locationName": "[parameters('locationName')]",
                        "isZoneRedundant": "[parameters('isZoneRedundant')]"
                    }
                ],
                "backupPolicy": {
                    "type": "Periodic",
                    "periodicModeProperties": {
                        "backupIntervalInMinutes": 240,
                        "backupRetentionIntervalInHours": 8
                    }
                },
                "isVirtualNetworkFilterEnabled": true,
                "virtualNetworkRules": [
                    {
                        "id": "redacted"
                    }
                ],
                "ipRules": ["redacted"],
                "dependsOn": [],
                "capabilities": [
                    {
                        "name": "EnableMongo"
                    },
                    {
                        "name": "EnableServerless"
                    }
                ],
<b>                "apiProperties": {
                    "serverVersion": "4.0" 
                },
</b>                "enableFreeTier": false
            },
            "tags": {
                "defaultExperience": "[parameters('defaultExperience')]",
                "hidden-cosmos-mmspecial": "",
                "CosmosAccountType": "Non-Production"
            }
        }
    ],
    "outputs": {}
}
</pre>

### New or Affected Resource(s)

* azurerm_cosmosdb_account

### Potential Terraform Configuration

```hcl
resource "azurerm_cosmosdb_account" "account" {
  name                = "myComsosDB"
  resource_group_name = "myGroup"
  location            = "East US"
  offer_type          = "Standard"
  kind                = "MongoDB"
  tags                = var.tags

  enable_automatic_failover = true

  capabilities {
    name = "EnableServerless"
  }

  api_properties {
    serverVersion = "4.0"
  }
}
```

**Comment by pricehatfield** [comment]
After doing some CLI testing, the ``--server-version`` option is apparently still in preview, so moving to that format may be premature.  

Upgrading an instance of Cosmos DB and importing it to the state file yields this on the next run:
```
Error: expected capabilities.1.name to be one of [EnableAggregationPipeline EnableCassandra EnableGremlin EnableTable EnableServerless EnableMongo MongoDBv3.4 mongoEnableDocLevelTTL DisableRateLimitingResponses AllowSelfServeUpgradeToMongo36 EnableAnalyticalStorage], got MongoDBv4.0
```

Implementing the MongoDBv4.0 option into the "capabilities" list would at least enable the newer version.

**Comment by evmimagina** [comment]
Hi,

Any ETA on this?

Thanks,

**Comment by derSchtefan** [comment]
The only workaround currently is to create the resource with TF and then run a script to migrate from 3.6 to 4.0 with the az command. Quite a hassle, we are relying on the 4.0 features but can't use TF to create our infrastructure. Implementation would be great, and I would opt for the version with:

```hcl
  capabilities {
    name = "EnableMongo"
  }
 api_properties {
    serverVersion = "4.0"
  }
```

As a reference: `az cosmosdb show ...`  displays this for 4.0 dbs:
```json
 "apiProperties": {
    "serverVersion": "4.0"
  },
  "capabilities": [
    {
      "name": "EnableMongo"
    },
    {
      "name": "EnableServerless"
    },
    {
      "name": "DisableRateLimitingResponses"
    }
  ],
```



For everybody who needs the workaround, execute this script after the terraform:

`az cosmosdb update --server-version 4.0 -g XXXXXX -n ZZZZZZ`

**Comment by katbyte** [comment]
Looks like this was fixed by #13757 - as such going to close this out, if it continues to be an issue please do open a new issue!
